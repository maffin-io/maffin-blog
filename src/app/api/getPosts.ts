import { DateTime } from 'luxon';
import { Octokit } from 'octokit';

import markdownToHtml from '@/lib/markdownToHtml';

export type Post = {
  slug: string,
  date: DateTime,
  title: string,
  summary: string,
  reading_time: string,
  tags: string[],
  content: string,
  author: {
    name?: string,
    avatar?: string,
  },
};

export async function getPosts(): Promise<Post[]> {
  const octokit = new Octokit();
  const issues = (await octokit.rest.issues.listForRepo({
    owner: 'maffin-io',
    repo: 'maffin-blog',
    creator: 'argaen',
    headers: {
      accept: 'application/vnd.github+json',
    },
  })).data.filter(issue => !issue.pull_request && !issue.title.includes('[draft]'));

  return Promise.all(issues.map(async (issue) => {
    if (!issue.body) {
      throw new Error(`Missing body in issue ${issue.title}`);
    }

    const { content: htmlContent, metadata } = await markdownToHtml(issue.body);
    return {
      slug: metadata.slug,
      date: DateTime.fromISO(issue.created_at),
      title: issue.title,
      summary: metadata.summary,
      reading_time: metadata.reading_time,
      // @ts-ignore
      tags: issue.labels.map(label => label.name),
      content: htmlContent,
      author: {
        name: issue.user?.name || issue.user?.login,
        avatar: issue.user?.avatar_url,
      },
    };
  }));
}

export async function getPost(slug: string): Promise<null | Post> {
  const posts = await getPosts();
  return posts.find(post => post.slug === slug) || null;
}

export async function getPostsByTag(tag: string): Promise<Post[]> {
  const posts = await getPosts();
  return posts.filter(post => post.tags.includes(tag));
}
