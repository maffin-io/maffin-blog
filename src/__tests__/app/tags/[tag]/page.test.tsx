import { DateTime } from 'luxon';
import { render } from '@testing-library/react';

import TagPage, { generateStaticParams } from '@/app/tags/[tag]/page';
import * as postsApi from '@/app/api/getPosts';
import type { Post } from '@/app/api/getPosts';

jest.mock('@/app/api/getPosts', () => ({
  __esModule: true,
  ...jest.requireActual('@/app/api/getPosts'),
}));

jest.mock('@/components/PostsList', () => {
  function PostsList({ posts }: { posts: Post[] }) {
    return (
      <div className="PostsList" data-props={`with-${posts.length}-posts`} />
    );
  }

  return PostsList;
});

async function resolveComponent(Component: React.FunctionComponent, props: any) {
  const ComponentResolved = await Component(props);
  return () => ComponentResolved;
}

describe('TagPage', () => {
  beforeEach(() => {
    jest.spyOn(postsApi, 'getPostsByTag').mockResolvedValue([]);
  });

  it('renders post with content and title', async () => {
    jest.spyOn(postsApi, 'getPostsByTag').mockResolvedValue([
      {
        title: 'My Blog Post 1',
      } as Post,
      {
        title: 'My Blog Post 2',
      } as Post,
    ]);
    const Component = await resolveComponent(TagPage, { params: { tag: 'label1' } });
    const { container } = render(<Component />);

    expect(container).toMatchSnapshot();
  });

  it('renders with no posts', async () => {
    const Component = await resolveComponent(TagPage, { params: { tag: 'label1' } });
    const { container } = render(<Component />);

    expect(container).toMatchSnapshot();
  });
});

describe('generateStaticParams', () => {
  it('returns one slug per post', async () => {
    jest.spyOn(postsApi, 'getPosts').mockResolvedValue([
      {
        content: '<div>Content</div>',
        slug: 'my-blog-post-1',
        summary: 'Blog post summary TODO',
        tags: ['label1', 'label2'],
        reading_time: '1 min',
        title: 'My Blog Post 1',
        date: DateTime.fromISO('2023-01-01'),
        author: {
          name: 'Name',
          avatar: 'https://avatar.image',
        },
      },
      {
        content: '<div>Content</div>',
        slug: 'my-blog-post-2',
        summary: 'Blog post summary TODO',
        tags: ['label2'],
        reading_time: '1 min',
        title: 'My Blog Post 2',
        date: DateTime.fromISO('2023-01-01'),
        author: {
          name: 'Name',
          avatar: 'https://avatar.image',
        },
      },
    ]);

    const staticParams = await generateStaticParams();

    expect(staticParams).toEqual([
      {
        tag: 'label1',
      },
      {
        tag: 'label2',
      },
    ]);
  });
});
