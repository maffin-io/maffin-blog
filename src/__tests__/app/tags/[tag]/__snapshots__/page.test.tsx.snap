// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagPage renders post with content and title 1`] = `
<div>
  <div
    class="divide-y divide-gray-200"
  >
    <div
      class="space-y-2 pt-6 pb-8 md:space-y-5"
    >
      <h1
        class="text-3xl font-extrabold leading-9 tracking-tight text-gray-900 sm:text-4xl sm:leading-10 md:text-6xl md:leading-14"
      >
        <a
          href="/tag/{tag}"
        >
          #
          label1
        </a>
      </h1>
    </div>
    <div
      class="PostsList"
      data-props="with-2-posts"
    />
  </div>
</div>
`;

exports[`TagPage renders with no posts 1`] = `
<div>
  <div
    class="divide-y divide-gray-200"
  >
    <div
      class="space-y-2 pt-6 pb-8 md:space-y-5"
    >
      <h1
        class="text-3xl font-extrabold leading-9 tracking-tight text-gray-900 sm:text-4xl sm:leading-10 md:text-6xl md:leading-14"
      >
        <a
          href="/tag/{tag}"
        >
          #
          label1
        </a>
      </h1>
    </div>
    <div
      class="PostsList"
      data-props="with-0-posts"
    />
  </div>
</div>
`;
